File: src/web/web.lua
local drystal = require 'drystal'
local web = drystal

web.raw_wget = web.wget
local wget_requests = {}

function drystal.on_wget_success(file)
	for _, f in ipairs(wget_requests[file].onload) do
		f(file)
	end
	wget_requests[file] = nil
end
function drystal.on_wget_error(file)
	for _, f in ipairs(wget_requests[file].onerror) do
		f(file)
	end
	wget_requests[file] = nil
end

function web.wget(url, file, onload, onerror)
	assert(onload)
	assert(onerror)
	if wget_requests[file] then
		-- already requested
		table.insert(wget_requests[file].onload, onload)
		table.insert(wget_requests[file].onerror, onerror)
		return
	end
	if not web.raw_wget(url, file) then
		return
	end
	wget_requests[file] ={
		onload={onload},
		onerror={onerror},
	}
end

File: src/web/web.c
/**
 * This file is part of Drystal.
 *
 * Drystal is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Drystal is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with Drystal.  If not, see <http://www.gnu.org/licenses/>.
 */
#ifdef EMSCRIPTEN
#include <assert.h>
#include <string.h>
#include <emscripten.h>

#include "web.h"
#include "dlua.h"
#include "lua_util.h"

char *run_js(const char* script)
{
	char *ret = emscripten_run_script_string(script);
	return strdup(ret);
}

static void onsuccess(const char* filename)
{
	assert(filename);

	if (dlua_get_function("on_wget_success")) {
		lua_State* L = dlua_get_lua_state();
		lua_pushstring(L, filename);
		CALL(1, 0);
	}
}

static void onerror(const char* filename)
{
	assert(filename);

	if (dlua_get_function("on_wget_error")) {
		lua_State* L = dlua_get_lua_state();
		lua_pushstring(L, filename);
		CALL(1, 0);
	}
}

void wget(const char* url, const char* filename)
{
	emscripten_async_wget(url, filename, onsuccess, onerror);
}
#endif

